Managers/EnemyManager.gd

```diff
@@ -2,7 +2,122 @@ extends Node
 ## Globally accessible for easy access to the enemies
 ## They are for now only accessible through the battler which is not very practical
 
-## The list of enemies to spawn
-@export var enemy_list: Array[Entity]:
-	set(list):
-		enemy_list = list
+## Distribution of enemies over the floors [br]
+## The values are arrays of PackedScene [br]
+## The keys are of format (X,Y,Z) with X the tree section from 1 to 3, 
+## Y the lower percent at which the enemy groups of the corresponding value are found,
+## Z the higher percent at which the enemy groups of the corresponding vlaue are found [br]
+## The intervals defined by the values are continuous, successive and non-overlapping.
+@export var enemy_group_distribution: Dictionary = {
+	[1, 0, 20]: [], # an array of packed scene for enemy groups
+	[1, 20, 40]: [],
+	[1, 40, 60]: [],
+	[1, 60, 80]: [],
+	[1, 80, 100]: [],
+	[2, 0, 20]: [], # section 2
+	[2, 20, 40]: [],
+	[2, 40, 60]: [],
+	[2, 60, 80]: [],
+	[2, 80, 100]: [],
+	[3, 0, 20]: [], # section 3
+	[3, 20, 40]: [],
+	[3, 40, 60]: [],
+	[3, 60, 80]: [],
+	[3, 80, 100]: [],
+}
+
+## The group of enemies currently being used, set at the start of the combat [br]
+var current_enemy_group: EnemyGroup
+
+## The packed scene corresponding to the group of enemies, this is used to keep track of it between saves [br]
+## DO NOT FORGET to set this back to null once the fight is done [br]
+## The reset is handled via signals to be sure to not forget it when changing the battler
+var current_enemy_group_packed: PackedScene = null
+
+## The last recorded lower bound of the current sub-section
+var current_lower_bound: int = enemy_group_distribution.keys()[0][1]
+## The last recorded higher bound of the current sub-section
+var current_higher_bound: int = enemy_group_distribution.keys()[0][2]
+## The last used array of enemy group
+var enemy_group_array: Array[PackedScene]
+## The shuffled version of the `enemy_group_array`
+var shuffled_enemy_group_array: Array[PackedScene] = []
+
+func _ready() -> void:
+	if enemy_group_array.is_empty():
+		enemy_group_array.assign(enemy_group_distribution.values()[0])
+	PhaseManager.on_event_win.connect(_clear_data)
+	PhaseManager.on_defeat.connect(_clear_data)
+
+
+## This calls the inner function that actually chooses the enemy group [br]
+## This is done to save the packed scene in this EnemyManger before it is returned
+func choose_enemy_group() -> PackedScene:
+	var packed_enemy_group: PackedScene = _choose_enemy_group_inner()
+	current_enemy_group_packed = packed_enemy_group
+	return packed_enemy_group
+
+## Chooses an enemy group in the following way: [br]
+## - Check if the player is in the same sub-section as the previous time this function was called [br]
+## - Choose an array of enemy group depending on player position if either sub-section changed, or the array to choose from is empty [br]
+## - Shuffle it, pop the last element of that array
+func _choose_enemy_group_inner() -> PackedScene:
```

Comment:
Loading of the game is currently broken - if you exit the game on a combat node and try to reload it crashes 
(1. enter combat, 2. exit game completely, 3. start again, 4. continue game -> Cannot call instantiate on a null object)
The current_enemy_group gets loaded, but current_enemy_group_packed is null and it gets returned from the function

Reply:
Fixed with 6f33aab92219